### PR TITLE
[SYNPY-1167] synapseutils.describe()

### PR DESCRIPTION
**Pull request by mattfazza**
## What's Inside

[This ticket](https://sagebionetworks.jira.com/browse/SYNPY-1167) has a lot of information on this issue.  This introduces a new function to synapseutils - unit tests included.

### CODE REVIEW

**Comment by mattfazza** [comment]
Will address the PEP8 points.

File: synapseutils/__init__.py [hunk 1]
```diff
@@ -14,10 +14,12 @@
 - :py:func:`monitor.notifyMe`
 - :py:func:`migrate_functions.index_files_for_migration`
 - :py:func:`migrate_functions.migrate_indexed_files`
+- :py:func:`describe.describe`
```
Comment (thomasyu888) [inline synapseutils/__init__.py:17]:
```suggestion
- :py:func:`describe_functions.describe`
```

File: synapseutils/describe_functions.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import synapseclient
+import pandas as pd
+from os import path
+from collections import defaultdict
+from json import dumps
```
Comment (thomasyu888) [inline synapseutils/describe_functions.py:5]:
```suggestion
from collections import defaultdict
from json import dumps
from os import path

import synapseclient
import pandas as pd
```

File: tests/unit/synapseutils/unit_test_synapseutils_describe.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import pytest
+import synapseclient
+import pandas as pd
+from numpy import array_equal
+import tempfile
+from synapseutils.describe_functions import _open_entity_as_df, _describe_wrapper
+from unittest.mock import patch
```
Comment (thomasyu888) [inline tests/unit/synapseutils/unit_test_synapseutils_describe.py:7]:
```suggestion
import tempfile
from unittest.mock import patch

from numpy import array_equal
import pandas as pd
import pytest
import synapseclient
from synapseutils.describe_functions import _open_entity_as_df, _describe_wrapper
```

File: synapseutils/describe_functions.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
```
Comment (thomasyu888) [inline synapseutils/describe_functions.py:2]:
```suggestion
import json
```

File: synapseutils/describe_functions.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+from os import path
```
Comment (thomasyu888) [inline synapseutils/describe_functions.py:3]:
```suggestion
import json
import os
```

Comment (thomasyu888) [reply synapseutils/describe_functions.py:3]: The rest of the client doesn't import particular functions, so I'm going to stick with that paradigm

File: synapseutils/describe_functions.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    :param syn: synapse object
+    :param entity: a synapse entity to be extracted and converted into a dataframe
+    :return: a pandas DataFrame if flow of execution is successful; None if not.
+    """
+
+    dataset = None
+
+    try:
+        entity = syn.get(entity)
+        name, format = path.splitext(entity.path)
```
Comment (thomasyu888) [inline synapseutils/describe_functions.py:21]:
```suggestion
        name, format = os.path.splitext(entity.path)
```

File: synapseutils/describe_functions.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                stats[column]['dtype'] = df[column].dtype
+            else:
+                stats[column]['mode'] = df[column].mode()[0]
+                stats[column]['dtype'] = df[column].dtype
+
+        except TypeError:
+            print("Invalid column type.")
+
+    if mode == 'string':
+        print(dumps(stats, indent=2, default=str))
```
Comment (thomasyu888) [inline synapseutils/describe_functions.py:63]:
```suggestion
        print(json.dumps(stats, indent=2, default=str))
```

**Review by thomasyu888 — APPROVED** [review]

**Comment by thomasyu888** [comment]
Thanks @mattfazza for the contribution :)

**Comment by mattfazza** [comment]
Thanks for the reviews!

**Pull request merged**